nemo_skills/inference/generate_solutions.py

```diff
@@ -133,29 +134,50 @@ def generate_solutions(cfg: GenerateSolutionsConfig):
 
             data_points.append(data_point)
 
-            if len(data_points) == cfg.batch_size:
-                # batch-computing the outputs
-                outputs = llm.generate(
-                    prompts=[prompt.build_string(data_point) for data_point in data_points],
-                    stop_phrases=list(cfg.prompt.stop_phrases),
-                    **asdict(cfg.inference),
-                )
-
-                for output, original_data_point in zip(outputs, data_points):
-                    # to make it easier to follow up with evaluation and limit accidental errors, we are adding
-                    # all of the ground-truth data to the output file alongside the generated solutions
-                    output.update(original_data_point)
-                    if 'error_message' not in output:
-                        output['error_message'] = extract_error_message(output['generation'])
-
-                    if cfg.generation_key != "generation":
-                        output[cfg.generation_key] = output.pop("generation")
-
-                    fout.write(json.dumps(output) + "\n")
-                data_points = []
+            if len(data_points) == cfg.batch_size or idx == len(data) - 1:
+                if cfg.inference.use_batch_api:
+                    # Using batch API for generation
+                    # Accessing the model directly since CodeExecutionWrapper does not support batch_generate
+                    request_metadata = llm.model.batch_generate(
+                        prompts=[prompt.build_string(data_point) for data_point in data_points],
+                        tokens_to_generate=cfg.inference.tokens_to_generate,
+                    )
+                    with open(cfg.output_file, 'at', encoding='utf-8') as fout:
+                        for data_point in data_points:
+                            fout.write(json.dumps(data_point) + '\n')
+
+                    # saving the request id to be able to retrieve results when they are ready
+                    with open(cfg.output_file + '-batch-request-id', 'wt', encoding='utf-8') as batch_fout:
+                        batch_fout.write(json.dumps({'request_id': request_metadata.id}))
+                    LOG.info('Submitted batch evaluation request. Please wait for the results to be ready.')
+                    LOG.info('The current status and final results can be accessed through summarize_results.py')
```

Comment:
Since this logic is decoupled from evaluation (there are cases where we run generation, but there are no metrics to compute), how about we add a new script inside pipeline, called `check_batch_gen.py` or something like that, which you will run with the same jsonl file list and it will do the check and fuse the results back into the original files

Reply:
Do you mean to handle batch result getting in a `check_batch_gen.py` instead of `summarize_results.py`?

Reply:
Yes, and the summarize_results.py (or actually better the compute_metrics function) will call that script first

Reply:
Done